File: src/bin/0704-binary-search.py
# @leetup=info id=704 lang=python3 slug=binary-search



# @leetup=code


class Solution:
    def search(self, nums: list[int], target: int) -> int:

        left = 0
        right = len(nums)

        while left < right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid
        return -1


# @leetup=code

File: src/bin/0704-binary-search.cpp
// @leetup=info id=704 lang=cpp slug=binary-search

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  int search(std::vector<int>& nums, int target) {
    auto it = std::lower_bound(nums.begin(), nums.end(), target);
    if (it == nums.end() || *it != target) {
      return -1;
    }
    return it - nums.begin();
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
